### PR TITLE
[WIP] Refactor tags service

### LINKED ISSUE #2142: async tags

### PR DESCRIPTION
**Pull request by Jont828**
**What type of PR is this?**

**What this PR does / why we need it**: Although the tags service doesn't quite fit into the async and `ResourceSpecGetter` interfaces, we can still do some refactoring to clean up the code.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2142 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note

```

### CODE REVIEW

File: azure/services/tags/spec.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-10-01/resources"
+	"github.com/Azure/go-autorest/autorest/to"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+)
+
+// TagsSpec defines the specification for a set of tags.
+type TagsSpec struct {
+	Scope string
+	Tags  infrav1.Tags
+	// Annotation is the key which stores the last applied tags as value in JSON format.
```
Comment (CecileRobertMichon) [inline azure/services/tags/spec.go:29]: comment needs to be updated

**Comment by CecileRobertMichon** [comment]
@Jont828 I think this smaller refactor makes sense. Please comment once it's ready for review, right now it's still failing tests and is marked as WIP.

**Comment by Jont828** [comment]
Sounds good, I was holding off on the tests until we decide to go with this refactor approach. I'll go ahead and add tests soon.

**Comment by Jont828** [comment]
After discussing with @CecileRobertMichon and @jackfrancis I think we can close out this refactor since the `tagsChanged()` function is quite difficult to understand or change in a way that ensures it actually works.

**Pull request closed without merging**